[Core] defensives prepull fix

otocephaly opened the pull request:
Fixes bugs relating to available timestamp adjustments for prepulls and prepulls never used again duplicates

## Pull request type

- [x] This is a bugfix to existing functionality

## Pull request details

<!-- If this PR is for new functionality or a bugfix, please complete the section below.  For a patch bump, delete this comment and the section below.  The context for your change is important to help the reviewer understand what the change is supposed to do - please provide an appropriate link or description to help the review process. -->
- [x] This is in response to a discussion or thread on Discord: *[Link to thread or start of discussion]*
    - https://discord.com/channels/441414116914233364/470049935077933076/1275859622196150423
    - https://discord.com/channels/441414116914233364/470049935077933076/1277048869619765403
- [x] The goal of this PR is detailed below:
    - fix prepull shenanigans in defensives

## Testing / Validation

<!-- If this PR contains any new functionality or bugfixes, please include log links that reviewers can use to help verify your changes.  Reviewers may also find additional logs to check your PR with, but having initial logs is important to speed the review process, especially for corner cases or hard to trigger analysis flags.  If this PR is a patch bump with no changes or potency only changes, you can delete this section. -->
- [x] I used the log(s) listed below to develop and test this bugfix:
    Log 1 from Protraction duplication https://www.fflogs.com/reports/v2VzJLCMwrxA1mPF#fight=1&source=9
    Log 2 for missing available suggestions https://www.fflogs.com/reports/762kPgLXJDp1wCAR#fight=6&source=2
    Log 3 for weird shake it off https://www.fflogs.com/reports/7WVCf1XMdnzATHQy#fight=11&source=3

## Job Maintenance
- [x] I have collaborated with one or more job maintainers for the job(s) in this PR while developing it

## Screenshots for bug fix
Log 1
![image](https://github.com/user-attachments/assets/e7025015-fbf8-4043-a6b5-c4ec9b4fef5b)

Log 2
![image](https://github.com/user-attachments/assets/0c26742b-b53a-4b32-add7-d3d8f345dcaf)

Log 3
![image](https://github.com/user-attachments/assets/b0976c5f-ee29-4750-834d-ca4cab70bab4)

otocephaly committed:
[Core] fix prepull defensive shenanigans

Fixes bugs relating to available timestamp adjustments for prepulls and prepulls never used again duplicates

otocephaly committed:
[Core] Prepull and clamping stuff

Noticed an issue with get max uses based on prepull shenanigans
Also removed clamping on current epoch

otocephaly committed:
changelog update

changelog update

otocephaly committed:
oop changelog

oop should read what I type

otocephaly committed:
Merge branch 'xivanalysis:dawntrail' into core_defensives_prepull_feex

otocephaly committed:
[SGE] feex argument on getExpirationTime

this only seemed to be a thing in dev for some reason

AkairyuGestalter commented on src/parser/jobs/sge/modules/Gauge.tsx:
```diff
@@ -210,7 +210,7 @@ export class Gauge extends CoreGauge {
 	//#endregion
 
 	private onComplete() {
-		const addersgallLeniency = ADDERSGALL_TIME_REQUIRED / 2
```
I'm not seeing the issue you mentioned in the PR notes in production at the moment, and I don't think this change is accurate. Can we drop it?

otocephaly replied:
okie removed!
(it's been a while since I looked at this, but I think it was a dev only issue, so if you happen to see it again, at least it won't come as a surprise lol)

AkairyuGestalter reviewed (changes requested):
I don't think the SGE part is necessary, and we need to fix up the changelog to pass the linter, but otherwise this seems fine.

otocephaly committed:
[sge] remove change

remove change

otocephaly committed:
Merge branch 'dawntrail' into core_defensives_prepull_feex

otocephaly committed:
Merge branch 'core_defensives_prepull_feex' of https://github.com/otocephaly/xivanalysis into core_defensives_prepull_feex

otocephaly committed:
lint pls

otocephaly committed:
updated date to be more current oop

otocephaly commented:
Hi @AkairyuGestalter , test seems to have an issue with removing the clamps
what would you recommend?
![image](https://github.com/user-attachments/assets/3b207b40-bd69-49d5-a76f-f6e670a1b8eb)

AzariahAeroborn commented:
That test failure is occurring because removing the clamps on getEpochTimestamp breaks the logic of the test -- now the very first action in the test with a -Infinity timestamp isn't getting clamped to the pull time of the fight and as such it fails the test which expects that it should be clamped.  

Overall, from looking at the actual change being made in the Defensives module, I'm pretty sure we don't need to remove the clamps at all.  Instead of the proposed changes, can we just:
1) Remove the call to tryGetAdditionalUsageRow at lines 127-129 at the beginning of the table
2) Update the code for tryGetUsageRow to check if the first usage is pre-pull.  
3) If the first usage is not pre-pull, have tryGetUsageRow output this.tryGetAdditionalUsageRow(defensive) before the current output, then continue on with the rest of the output.
4) If the first usage is pre-pull, only output the current output (don't check for additional usages before the first use, since the first use was prepull)

For log #2, the issue may be coming from line 171 in the current production code -- that line seems to assume pre-pull events will have a timestamp _equal to_ the pull time, and that isn't true any more... the timestamp will be less than the pull timestamp.  Maybe just changing that to a <= check will fix it?

AzariahAeroborn reviewed (dismissed):
See comment in discussion thread.

otocephaly committed:
Merge branch 'dawntrail' into core_defensives_prepull_feex

otocephaly committed:
revert clamp back to normal and fix prepull logic

title

otocephaly commented:
Hi @AzariahAeroborn @AkairyuGestalter 

I reverted clamps back to normal as recommended (I wonder if we should clamp to the lowest expected prepull e.g. something like this.parser.pull.timestamp - 30 seconds just for display / calculation purposes. for now, I leave as is)

For the prepull in defensives stuff, I used the logic I had to remove the duplicates in the table and added it to the getAdditionalUsage since you recommended looking there. I think when I added that prepull available timestamp adjustment line way back when, I made it a lil too convoluted that I am not sure what I was thinking lol... anywho, I updated the former line 162 to include a check for prepull since it assumed if the timestamp was the same then it was available from the beginning. I left a comment to show my thought process on that line. I removed the old convoluted logic for whatever it was doing

Thank you

AkairyuGestalter reviewed (approved):
This is looking good now! And just in time for FRU to really make people think about defensives!
